src/components/NodeEditor/NodeEditor.tsx

```diff
@@ -78,16 +83,95 @@ export interface NodeEditorProps {
     onUpdateLocalizations: Function;
     onUpdateAction: Function;
     onUpdateRouter: Function;
-    /** Perform when editor is closed */
     onClose?(canceled: boolean): void;
     ComponentMap: ComponentMap;
 }
 
 export interface NodeEditorState {
     config: Type;
     show: boolean;
+    setResultName: boolean;
 }
 
+export const mapExits = (exits: Exit[]): { [uuid: string]: Exit } =>
+    exits.reduce(
+        (map, exit) => {
+            map[exit.uuid] = exit;
+            return map;
+        },
+        {} as { [uuid: string]: Exit }
+    );
+
+export const isSwitchForm = (type: string) =>
+    type === 'wait_for_response' || type === 'expression' || type === 'group';
+
+/**
+ * Returns existing action (if any), or a bare-bones representation of the form's action.
+ */
+export const getAction = ({ action }: NodeEditorProps, config: Type): AnyAction => {
+    let uuid: string;
+
+    if (action) {
+        if (action.type === config.type || (config.aliases && action.type === config.aliases[0])) {
+            return action;
+        } else {
+            ({ uuid } = action);
+        }
+    }
+
+    let defaultAction: Action = {
+        type: config.type,
+        uuid: uuid || generateUUID()
+    };
+
+    switch (config.type) {
+        case 'reply':
+            defaultAction = { ...defaultAction, text: '', all_urns: false } as Reply;
+            break;
+        case 'add_to_group':
+            defaultAction = { ...defaultAction, groups: null } as ChangeGroup;
+            break;
+        case 'save_contact_field':
+            defaultAction = {
+                ...defaultAction,
+                field_uuid: generateUUID(),
+                field_name: '',
+                value: ''
+            } as SaveToContact;
+            break;
+        case 'send_email':
+            defaultAction = { ...defaultAction, subject: '', body: '', emails: null } as SendEmail;
+            break;
+        case 'save_flow_result':
+            defaultAction = {
+                ...defaultAction,
+                result_name: '',
+                value: '',
+                category: ''
+            } as SaveFlowResult;
+            break;
+        case 'call_webhook':
+            defaultAction = { ...defaultAction, url: '', method: 'GET' } as CallWebhook;
+            break;
+        case 'start_flow':
+            defaultAction = { ...defaultAction, flow_name: null, flow_uuid: null } as StartFlow;
+            break;
+    }
+
+    return defaultAction;
+};
+
+export const FormContainer: React.SFC<{
```

Comment:
`FormContainer` should be declared outside of NodeEditor and it's `className` prop, which shadows React's `className` prop should be prefixed with a dunder to indicate it's a special case.